File: 0055-jump-game/0055-jump-game.cpp
#include <vector>
#include <algorithm>

class Solution {
public:
    bool canJump(vector<int>& nums) {
        int maxReach = 0;
        int n = nums.size();
        
        for (int i = 0; i < n; ++i) {
            // If the current index is beyond the maximum index we can reach, fail
            if (i > maxReach) {
                return false;
            }
            
            // Update the furthest index we can reach from here
            maxReach = max(maxReach, i + nums[i]);
            
            // Early exit if we can already reach or exceed the last index
            if (maxReach >= n - 1) {
                return true;
            }
        }
        
        return true;
    }
};

File: 0055-jump-game/0055-jump-game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        fp = len(nums) - 1
        
        for p in range(len(nums) - 2, -1, -1):
            # Check if current index (p) + jump capability (nums[p]) reaches the target (fp)
            if p + nums[p] >= fp:
                fp = p 
                
        return fp == 0